src/world_interface/world_interface.h

```diff
@@ -1,16 +1,94 @@
+#pragma once
 
 #include "../simulator/utils.h"
 
-void world_interface_init(); // Call this before trying to do anything else
+#include <chrono>
+
+// the clock used for time measurements for data
+using dataclock = std::chrono::steady_clock;
+// a point in time as measured by dataclock
+using datatime_t = std::chrono::time_point<dataclock>;
+
+/**
+ * @brief Represents data measured using a sensor at a given time.
+ *
+ * @tparam T The type of data measured from the sensor. Requires defined default constructor.
+ */
+template <typename T> class DataPoint {
+public:
+    /**
+     * @brief Construct an invalid DataPoint, holding no data
+     */
+    DataPoint() : valid(false), time(), data() {}
+
+	/**
+	 * @brief Construct a new DataPoint object, measured now.
+	 *
+	 * @param data The piece of data
+	 */
+	DataPoint(T data) : DataPoint(dataclock::now(), data) {}
+
+	/**
+	 * @brief Construct a new DataPoint object, measured at the given time.
+	 *
+	 * @param time The time at which the data was measured.
+	 * @param data The piece of data.
+	 */
+	DataPoint(datatime_t time, T data) : valid(true), time(time), data(data) {}
+
+	// provide an implicit conversion to the data type. Helps with backwards compatability.
+	operator T() const { return data; }
+
+    // check if this object holds any data
+    operator bool() const { return valid; }
+
+    /**
+     * @brief Check if this measurement was taken recently.
+     * 
+     * @param duration The data is fresh if it was taken at most this many milliseconds ago.
+     * @return true if the data was measured at most \p duration milliseconds ago, false otherwise.
+     */
+    bool isFresh(std::chrono::milliseconds duration) {
+        return dataclock::now() - duration <= time;
+    }
+
+    // true iff this measurement is valid, false otherwise.
+    bool valid; 
+	// the time at which the data was measured. Use only if valid == true.
+	datatime_t time;
+	// the measurement data. Use only if valid == true.
+	T data;
```

Comment:
I know this class will only be used internally by us, but it would probably be best if these were private fields with accessor methods just to make sure we can enforce class invariants (e.g. valid should not be `true` if there's no data).

Reply:
Makes sense!